### PR TITLE
Improve Directory Group lookup error handling

### PR DESCRIPTION
**Pull request by ghsolomon**

### CODE REVIEW

File: src/main/groovy/io/xh/hoist/role/provided/DefaultRoleService.groovy [hunk 1]
```diff
@@ -269,18 +272,10 @@ class DefaultRoleService extends BaseRoleService {
...
-       if (!directoryGroups) return emptyMap()
-        // Wrapped here to avoid failing implementations from ever bringing down app.
-        try {
-            return doLoadUsersForDirectoryGroups(directoryGroups)
-        } catch (Throwable e) {
-            def errMsg = 'Error resolving Directory Groups'
-            logError(errMsg, e)
-            return directoryGroups.collectEntries {[it, errMsg] }
-        }
+        !directoryGroups ? emptyMap() : doLoadUsersForDirectoryGroups(directoryGroups)
```
Comment (lbwexler) [inline src/main/groovy/io/xh/hoist/role/provided/DefaultRoleService.groovy:275]: perhaps inline the "do" method -- one less method to trace through

**Review by lbwexler — APPROVED** [review]

File: grails-app/services/io/xh/hoist/role/provided/DefaultRoleAdminService.groovy [hunk 2]
```diff
@@ -41,9 +41,15 @@ class DefaultRoleAdminService extends BaseService {
         if (defaultRoleService.directoryGroupsSupported) {
             Set<String> groups = roles.collectMany(new HashSet()) { it.directoryGroups }
             if (groups) {
-                Map<String, Object> groupsLookup = defaultRoleService.loadUsersForDirectoryGroups(groups)
-                usersForGroups = groupsLookup.findAll { it.value instanceof Set }
-                errorsForGroups = groupsLookup.findAll { !(it.value instanceof Set) }
+                try {
+                    Map<String, Object> groupsLookup = defaultRoleService.loadUsersForDirectoryGroups(groups)
+                    usersForGroups = groupsLookup.findAll { it.value instanceof Set }
+                    errorsForGroups = groupsLookup.findAll { !(it.value instanceof Set) }
```
Comment (amcclain) [inline grails-app/services/io/xh/hoist/role/provided/DefaultRoleAdminService.groovy:47]: To confirm - we might still get `errorsForGroups` populated as expected with partial lookup failures, but now only in the case of 'No AD group found' - is that correct?  

Checking that a search returnign no result does not throw- that would be reasonable but didn't want to assume.

(Would suggest changing to "Directory Group not found" FWIW)

File: src/main/groovy/io/xh/hoist/role/provided/DefaultRoleService.groovy [hunk 3]
```diff
@@ -290,16 +285,22 @@ class DefaultRoleService extends BaseRoleService {
...
+                loadUsersForDirectoryGroups(groups).each { k, v ->
+                    if (v instanceof Set) {
+                        usersForDirectoryGroups[k] = v
+                    } else {
+                        logError("Error resolving users for directory group", k, v)
+                    }
                 }
+                _usersForDirectoryGroups = usersForDirectoryGroups
+            } catch (Throwable e) {
+                logError("Error resolving users for directory groups", e)
```
Comment (amcclain) [inline src/main/groovy/io/xh/hoist/role/provided/DefaultRoleService.groovy:303]: Worth a comment to clarify that leaving the previous `_usersForDirectoryGroups` in place (and then re-processing it below) is a deliberate choice?

**Comment by lbwexler** [comment]
Trying to think big picture here and wanted to get this out here.  In particular, don't want to trade one set of problems for another.

What motivated this change was the concern that we are making a large number of individual LDAP queries to get groups right now.  If any one of them fails (e.g. timeout, or just flaky) we can basically "blip out" user access for all of the users in the group that failed.  And we may be just doing our 10 minute refresh, replacing a perfectly good result with a flaky result -- and that seemed wrong.   So this should fix that.

New concern --  now a single flaky call keeps us from ever getting needed roles in the app. Or worse yet, someone enters something as a new or modified group tag that somehow causes all role groups to fail to load.

Maybe its overkill, but we could pass a flag to indicate whether we wanted "strict mode" or whether we are happy to have whatever results we can get?  The latter could be useful when we start-up, or for whatever reason have no successful load of groups.

File: grails-app/services/io/xh/hoist/role/provided/DefaultRoleAdminService.groovy (hunk 2 above)
Comment (ghsolomon) [reply grails-app/services/io/xh/hoist/role/provided/DefaultRoleAdminService.groovy:47]: Correct - will double-check this now to make sure. And sounds good re: change to error text

Comment (ghsolomon) [reply grails-app/services/io/xh/hoist/role/provided/DefaultRoleAdminService.groovy:47]: Confirmed! Built a snapshot from this branch and tested in an EMC app to confirm.

**Comment by lbwexler** [comment]
Created a new branch here to incorporate my thoughts above.  Greg reveiwed and testing out now

**Pull request closed without merging**